py/controller.py

```diff
@@ -0,0 +1,230 @@
+from colorama import Fore
+import visa
+
+# import gpib
+
+# initialize colorama
+colorama.init(autoreset=True)
+
+# latency of controller's reaction to user input and internal messages/signals
+LATENCY = 0.1
+
+class InstrumentController(mp.Process):
+    """Controller for a pyvisa instrument.
+
+    Inherited from multiprocessing.Process.
+
+    Attributes:
+    messenger - multiprocessing.Connection object for sending and receiving
+                messages to and from the instrument.
+    """
+
+    # Internal attributes:
+    # _instrument - visa.Resource object for the instrument
+    # _listener   - the other end of messenger's pipe
+
+    def __init__(self, instrument):
+        # if not isinstance(instrument, visa.Resource):
+        #     err_msg = 'instrument has type %s, not visa.Resource' % \
+        #               type(instrument).__name__
+        #     raise TypeError(err_msg, instrument)
+        mp.Process.__init__(self)
+        self._instrument = instrument
+        self.messenger, self._listener = mp.Pipe()
+
+class PowerSupplyController(InstrumentController):
+
+    def __init__(self):
+        InstrumentController.__init__(self, None)
+        # InstrumentController.__init__(self, gpib.PowerSupply())
+        # self._instrument.initialize()
+
+    def run(self):
+        pass
+
+class LockInController(InstrumentController):
+
+    def __init__(self):
+        InstrumentController.__init__(self, None)
+        # InstrumentController.__init__(self, gpib.LockIn())
+        pass
+
+    def run(self):
+        pass
+
+def console_control(verbose_prompt=True):
+    psc = PowerSupplyController()
+    lic = LockInController()
+    psc.start()
+    lic.start()
+    pscm = psc.messenger
+    licm = lic.messenger
+    long_prompt = """Please type in one of the following commands:
+   * r(amp) TARGET_CURRENT RAMPING_RATE
+        ramp up/down current output of the power supply
+   * i(nterrupt)
+        interrupt ramping (only useful when ramping is in progress)
+   * f(inish)
+        finish experiment and terminate controllers
+   * k(ill)
+        terminate controllers
+"""
+    short_prompt = "Commands: r(amp) TARGET_CURRENT RAMPING_RATE, i(nterrupt), f(inish), or k(ill)\n"
+    status = 0 # status of last command, 0 if successful, 1 if failed
+    while True:
+        # prompt for and read console input
+        if verbose_prompt:
+            sys.stderr.write(Fore.CYAN + long_prompt)
+        else:
+            sys.stderr.write(Fore.CYAN + short_prompt)
+        if status == 0:
+            sys.stderr.write(Fore.GREEN + ">> ")
+        else:
+            sys.stderr.write(Fore.RED + ">> ")
+        try:
+            console_input = raw_input()
+        except NameError:
+            console_input = input()
+
+        # consume all messages from instruments during the period of waiting
+        # might add error diagnostic and handling code here
+        while psc.messenger.poll():
+            psc.messenger.recv()
```

Comment:
Possible enhancement: send a `ramping completed` message from the power supply controller, and add processing code here. However, the downside is that `raw_input` or `input` (py3k) will always block, delaying the status update. Also, the `ramping completed` message will introduce race conditions unless we also add processing blocks in other places. This could be overcomplicated and we can just monitor the power supply display panel.